proposals/list-patterns.md

```diff
@@ -0,0 +1,142 @@
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is {1, 2, 3}` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+The pattern syntax is modified as follow:
+
+```antlr
+positional_pattern
+  : type? positional_pattern_clause length_pattern_clause? property_or_list_pattern_clause? simple_designation?
+  ;
+
+length_pattern
+  : type? length_pattern_clause property_or_list_pattern_clause? simple_designation?
+  ;
+
+list_pattern
+  : type? list_pattern_clause simple_designation?
+  ;
+
+property_pattern
+  : type? property_pattern_clause simple_designation?
+  ;
+
+length_pattern_clause
+  : '[' pattern ']'
+  ;
+
+property_or_list_pattern_clause
+  : list_pattern_clause
+  | property_pattern_clause
+  ;
+
+property_pattern_clause
+  : '{' (subpattern (',' subpattern)* ','?)? '}'
+  ;
+
+list_pattern_clause
+  : '{' pattern (',' pattern)* ','? '}'
+  ;
+
+slice_pattern
+  : '..' negated_pattern?
+  ;
+
+primary_pattern
+  : list_pattern
+  | length_pattern
+  | slice_pattern
+  | // all of the pattern forms previously defined
+  ;
+```
+There are three new patterns:
+
+- The *list_pattern* is used to match elements and the *length_pattern* is used to match the length.
+- A *slice_pattern* is only permitted once and only directly in a *list_pattern_clause* and discards _**zero or more**_ elements.
+
+Notes:
+
+- Due to the ambiguity with *property_pattern*, a *list_pattern* cannot be empty and a *length_pattern* should be used instead to match a list with the length of zero, e.g. `[0]`. 
+- The *length_pattern_clause* must be in agreement with the inferred length from the *list_pattern_clause* (if any), e.g. `[0] {1}` is an error.
+	- However, `[1] {}` is **not** an error due to the length mismatch, rather, `{}` would be always parsed as an empty *property_pattern_clause*. We may want to add a warning for it so it would not be confused that way.
+- If the *type* is an *array_type*, the *length_pattern_clause* is disambiguated so that `int[] [0]` would match an empty integer array.
+- All other combinations are valid, for instance `T (p0, p1) [p2] { name: p3 } v` or `T (p0, p1) [p2] { p3 } v` where each clause can be omitted.
+
+> **Open question**: Should we support all these combinations?
+
+#### Pattern compatibility
+
+A *length_pattern* is compatible with any type that is *countable*, i.e. has an accessible property getter that returns an `int` and has the name `Length` or `Count`. If both properties are present, the former is preferred.
+
+A *list_pattern* is compatible with any type that conforms to the following rules:
+
+1. Is compatible with the *length_pattern*
+2. Has an accessible indexer with a single `int` parameter
+
+ > **Open question**: Should we support `this[Index]` indexers? If so, which one is preferred if `this[int]` is also present?
```

Comment:
Started thread with LDM and @alrz. 
I think we should just say that:
- if the type is "indexable" (explicitly or implicitly, ie. can you write `expr[^1]`?) then the list pattern can be used
- if the type is "sliceable" (explicitly or implicitly, ie. can you write `expr[1..^1]`?) then the slice pattern can be used
- if the type is countable (defined in ranges spec) then the length pattern can be used

Reply:
That makes perfect sense (I assume this is the terminology we used for ranges). Following range semantics that would imply we prefer this[Index] and this[Range] if both are present.

Reply:
Quick question: Does this mean we may use `this[int]` for leading patterns and `this[Index]` for trailing patterns?

Reply:
Confirmed with LDM that we should use this approach.
The ranges spec only defined "countable". But it unfortunately doesn't define "indexable" or "sliceable" as concepts. It uses terms like "Index support" and "Range support". I think that's okay, we can introduce the terminology now.

> However, if we were to completely match indexer access, does that mean we may use this[int] for leading patterns and this[Index] for trailing patterns?

Good point. Both the leading and trailing patterns should bind the same way.
We can solve this by saying that we bind like `expr[new Index(index)]`.